src/vs/editor/contrib/suggest/browser/suggestController.ts

```diff
@@ -558,12 +560,54 @@ export class SuggestController implements IEditorContribution {
 			resolveInfo: !item.provider.resolveCompletionItem ? -1 : itemResolved ? 1 : 0,
 			resolveDuration: item.resolveDuration,
 			commandDuration: commandExectionDuration,
-			additionalEditsAsync: additionalEditsAppliedAsync,
-			index,
-			label: item.textLabel
+			additionalEditsAsync: additionalEditsAppliedAsync
 		});
 	}
 
+	private _reportSuggestionIndexTelemetry(item: CompletionItem, index: number) {
+		if (this.completionItems === undefined) {
+			return;
+		}
+
+		const selectedLabel = item.textLabel;
+
+		// Grab first 30 sorted suggestions
+		const completionItems = this.completionItems?.splice(0, 30);
+
+		// Determine if we have duplicates in the list
+		const hasOtherInstances = (items: CompletionItem[], label: string, index: number): boolean => {
+			return items.some((item, i) => i !== index && item.textLabel === label);
+		};
+
+		const hasDuplicates = hasOtherInstances(completionItems, selectedLabel, index);
+
+		// If duplicate, find first duplicate in the list
+		const firstIndex = hasDuplicates ? completionItems.findIndex((completion) => completion.textLabel === selectedLabel) : -1;
+
```

Comment:
I think you improve performance here because an array is allocated (`splice`) and later that array is looped over two times. Can't you have a for-loop (till 30 or item length) that populates a map that is keyed by label and contains indexes? With that a simple look you give the first duplicate (if any).

Reply:
Also note the danger of using (destructive) `splice` on a property. The property should not exist (see above) but splice would alter it and potentially trigger weird effects.